Pal 872 user service streaming documentation (#276)
gchq/Palisade-services: user-service/README.md

The User Service separates this concern from the rest of the system. Other components use this service's API to request user details. 
Some deployments may also allow Palisade to add users to the system, hence the presence of the `addUser()` method in the `UserService` interface.

## High Level Architecture

<!--- 
See user-service/doc/user-service.drawio for the source of this diagram
--->
![User Service diagram](doc/user-service.png)

> ## Message Model and Database Domain

Reviewer: ```suggestion
## Message Model
```

The User Service separates this concern from the rest of the system. Other components use this service's API to request user details. 
Some deployments may also allow Palisade to add users to the system, hence the presence of the `addUser()` method in the `UserService` interface.

## High Level Architecture

<!--- 
See user-service/doc/user-service.drawio for the source of this diagram
--->

> ![User Service diagram](doc/user-service.png)

Reviewer: Can we change persistence to cache?

Reviewer: The kafka bits are all better described as topics, then document usage with a verb like:
"Source from kafka topic 'request'"
"Sink to kafka topic 'user'"
"Sink to kafka topic 'error'"

Author: The diagram has been updated with the new descriptions

## Message Model and Database Domain

> | UserRequest     | UserResponse     | AuditErrorMessage | AuditableUserResponse |

Reviewer: I think you could leave out the AuditableUserResponse, it's just used as an internal 'pair' and doesn't belong in this level of documentation

Author: Ok, removed the AuditableUserResponse details.

}'
```

## Uploading users to the backing store on service start-up

> It may be that some example users needs to be added to the backing store before a test run of the Palisade system is performed, for example. This is resolved by using Spring to upload the user(s) to the service from a yaml file. 

Reviewer 2: I dont really like the use of 'upload' here, as its not really what the service is doing. It would be better to mention cache warming, and inserting into a flavour of the cache layer, depending on profiles

Author: Updated

}'
```

## User pre-population / Cache Warming

This service can be populated with a list of initial users at runtime.
By specifying all the configuration and pre-population factory values, any user information in the configuration list will be added to the service at runtime.
This includes being added to the cache, so that the basic `NullUserService` implementation, coupled with the `UserServiceCachingProxy`, acts solely using the cache.

> **Note:** This means that after the cache eviction TTL (Time To Live), the User Service will no longer contain any user information.

Reviewer 2: add where it can be found:
```
**Note:** This means that after the cache eviction TTL (Time To Live), found in [application.yaml(application.yaml) the User Service will no longer contain any user information.
```

Author: Updated

> An example of this can be seen in this [Test Yaml](src/contract-tests/resources/application-pre-population.yaml) file which adds the basic user information to the backing store when the service starts up.

Reviewer 2: ```suggestion
An example of this can be seen in this [Test Yaml](src/contract-tests/resources/application-pre-population.yaml) file which adds a basic User to the backing store when the service starts up.
```

}'
```

## User pre-population / Cache Warming

> This service can be populated with a list of initial users at runtime.

Reviewer 3: bit confusing, not "at runtime" but "at initialisation" or "at start-up".

Author: Amended this to read "at startup"

By specifying all the configuration and pre-population factory values, any user information in the configuration list will be added to the service at runtime.
This includes being added to the cache, so that the basic `NullUserService` implementation, coupled with the `UserServiceCachingProxy`, acts solely using the cache.

**Note:** This means that after the cache eviction TTL (Time To Live), found in [application.yaml](src/main/resources/application.yaml), the User Service will no longer contain any user information.
 

> An example of this can be seen in this [Test Yaml](src/contract-tests/resources/application-pre-population.yaml) file which adds a basic user to the backing store when the service starts up.

Reviewer: ```suggestion
An example of this can be seen in this [Test Yaml](src/contract-tests/resources/application-pre-population.yaml) file which adds a basic user to the cache when the service starts up.
```

Author: Updated
